binutils-unwrapped, gcc, gdb: force targetPrefix if cross compiling

sternenseemann opened the pull request:
The binutils build system checks by itself if it is building a cross
toolchain or not and prepends or omits a targetPrefix accordingly. This
means that we can always pass target via configureTargets.

However the binutils build system and our bintools wrapper disagree over
whether we are building a cross toolchain or not sometimes since cross
compilation can be relatively subtle in nixpkgs. For example every use
of crossOverlays will make nixpkgs build a cross toolchain even though
localSystem == crossSystem. The cross infrastructure is also used to
build native binaries with a different stdenv (musl instead of glibc,
clang instead of gcc). In all of these cases stdenv.hostPlatform.config
== stdenv.targetPlatform.config, causing binutils to not prepend a
target prefix. At the same time stdenv.hostPlatform !=
stdenv.targetPlatform causing the bintools wrapper to expect a target
prefix, thus building an incomplete set of bintools. This is why
currently pkgsCross.gnu64 and pkgsCross.musl64 aren't working.

The solution is quite simple however: If we detect that we are building
a cross toolchain in the binutils-unwrapped expression, we force the
targetPrefix with --programprefix and fulfill the expectations of the
bintools wrapper at the same time.

Tested (on x86_64-linux):

* pkgsCross.musl64.hello
* pkgsCross.aarch64-multiplatform.hello
* pkgs.hello

Still not working is pkgsCross.gnu64.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

sternenseemann committed:
binutils-unwrapped: force targetPrefix if cross compiling

The binutils build system checks by itself if it is building a cross
toolchain or not and prepends or omits a targetPrefix accordingly. This
means that we can always pass target via configureTargets.

However the binutils build system and our bintools wrapper disagree over
whether we are building a cross toolchain or not sometimes since cross
compilation can be relatively subtle in nixpkgs. For example every use
of crossOverlays will make nixpkgs build a cross toolchain even though
localSystem == crossSystem. The cross infrastructure is also used to
build native binaries with a different stdenv (musl instead of glibc,
clang instead of gcc). In all of these cases stdenv.hostPlatform.config
== stdenv.targetPlatform.config, causing binutils to not prepend a
target prefix. At the same time stdenv.hostPlatform !=
stdenv.targetPlatform causing the bintools wrapper to expect a target
prefix, thus building an incomplete set of bintools. This is why
currently pkgsCross.gnu64 and pkgsCross.musl64 aren't working.

The solution is quite simple however: If we detect that we are building
a cross toolchain in the binutils-unwrapped expression, we force the
targetPrefix with --programprefix and fulfill the expectations of the
bintools wrapper at the same time.

Tested (on x86_64-linux):

* pkgsCross.musl64.hello
* pkgsCross.aarch64-multiplatform.hello
* pkgs.hello

Still not working is pkgsCross.gnu64, since
x86_64-unknown-linux-gnu-stage-final-gcc gets confused about targets
now, so bootstrapping the stdenv fails. Since this wasn't working
previously anyways, it's proably fine to fix this separately.

sternenseemann commented:
For reference, this is the new failure when building `pkgsCross.gnu64.hello` on x86_64-linux:

```
*** This configuration is not supported in the following subdirectories:
     gnattools gotools target-libada target-libhsail-rt target-libphobos target-libbacktrace target-libgfortran target-libgo target-libffi target-libobjc target-liboffloadmic
    (Any other directories should still work fine.)
*** --with-headers is only supported when cross compiling
builder for '/nix/store/jn1jnbb6wabgxhiafdbwbk809gag1nv3-x86_64-unknown-linux-gnu-stage-final-gcc-10.2.0.drv' failed with exit code 1
error: build of '/nix/store/jn1jnbb6wabgxhiafdbwbk809gag1nv3-x86_64-unknown-linux-gnu-stage-final-gcc-10.2.0.drv' on 'ssh://edwin' failed: builder for '/nix/store/jn1jnbb6wabgxhiafdbwbk809gag1nv3-x86_64-unknown-linux-gnu-stage-final-gcc-10.2.0.drv' failed with exit code 1
builder for '/nix/store/jn1jnbb6wabgxhiafdbwbk809gag1nv3-x86_64-unknown-linux-gnu-stage-final-gcc-10.2.0.drv' failed with exit code 1
cannot build derivation '/nix/store/ri3gpxdij4pm5x6cljzkxa9qrdawxqdz-x86_64-unknown-linux-gnu-stage-final-gcc-wrapper-10.2.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/d2dyi5c0zqbaifz4r81sc9az78nmsk9p-stdenv-linux.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/7c8l5ikhlwkhlysf0j1byrwkj4h1s3w4-hello-2.10-x86_64-unknown-linux-gnu.drv': 1 dependencies couldn't be built
error: build of '/nix/store/7c8l5ikhlwkhlysf0j1byrwkj4h1s3w4-hello-2.10-x86_64-unknown-linux-gnu.drv' failed
```

Ericson2314 commented on pkgs/development/tools/misc/binutils/default.nix:
```diff
@@ -126,7 +125,16 @@ stdenv.mkDerivation {
     # RUNPATH can be overriden using LD_LIBRARY_PATH at runtime.
     "--enable-new-dtags"
-  ] ++ lib.optionals gold [ "--enable-gold" "--enable-plugins" ];
+  ] ++ lib.optionals gold [
+    "--enable-gold"
+    "--enable-plugins"
+  ] ++ lib.optionals (stdenv.hostPlatform != stdenv.targetPlatform) [
+    # force target prefix if we are cross-compiling since binutils
+    # doesn't set the prefix if stdenv.hostPlatform.config and
+    # stdenv.targetPlatform.config are the same which may happen
+    # even though we are cross compiling (e. g. pkgsCross.musl64).
+    "--program-prefix=${targetPrefix}"
+  ];
```
```suggestion
    # force target prefix. Some versions of binutils will make it empty
    # if `--host` and `--target` are too close, even if Nixpkgs thinks the
    # platforms different (e.g. because not all the info makes the `config`).
    # Other versions of binutils will always prefix if `--target` is passed, even
    # if `--host` and `--target`are the same. The easiest thing for us to do
    # is not leave it to chance, and force the program prefix to be what we
    # want it to be.
    "--program-prefix=${targetPrefix}"
  ] ++ lib.optionals gold [
    "--enable-gold"
    "--enable-plugins"
  ];
```

sternenseemann replied:
Done, thank you!

sternenseemann committed:
gcc*: always force --program-prefix and pass --target

Don't rely on gcc detecting from the passed platforms which prefix to
use, but always specify the prefix nixpkgs expects (or doesn't). This
allows us to work around problems where the configure script would add
prefix where nixpkgs doesn't expect one (if `--target` was specified,
but the same as `--host`) or doesn't add one if nixpkgs expects one (if
`--target` and `--host` are the same, but we are actually cross
compiling, but the relevant parts of the platform are not encoded into
the platform config.

See also ca9be0511b0661be5255dd1fe0b05c5eaf3e56d8.

sternenseemann committed:
gdb: always set --program-prefix and pass --target

This makes the targetPrefix always conform to nixpkgs' expectations
instead of relying on the autotools build system to figure it out
correctly (which is also inconsistent across versions).

See also ca9be0511b0661be5255dd1fe0b05c5eaf3e56d8.

Ericson2314 reviewed (approved):
Looks great!

Ericson2314 commented:
I built `pkgsCross.aarch64-multiplatform.buildPackages.gdb`. This stuff looks good to me.

Ericson2314 commented:
`pkgsCross.aarch64-multiplatform.hello` too.

sternenseemann commented:
Great! I've also checked: All `gcc` versions in `pkgs` build, stuff in `pkgsCross.musl64` compiles including `gdb` and also `pkgsCross.aarch64-multiplatform.buildPackages.gdb` :)
